executor/src/witgen/rows.rs

```diff
@@ -195,9 +201,63 @@ impl<T: FieldElement> Debug for Cell<'_, T> {
     }
 }
 
-/// A row of cells, indexed by polynomial ID.
 pub type Row<'a, T> = WitnessColumnMap<Cell<'a, T>>;
 
+/// Merges two rows, updating the first.
+/// Range constrainst from the second row are ignored.
+/// @returns false if there was a conflict (two different known values)
+#[must_use]
+pub fn merge_row_with<'a, T: FieldElement>(row: &mut Row<'a, T>, other: &Row<'a, T>) -> bool {
+    // First check for conflicts.
+    if row
+        .values()
+        .zip(other.values())
+        .any(|(cell1, cell2)| match (&cell1.value, &cell2.value) {
+            (CellValue::Known(v1), CellValue::Known(v2)) => v1 != v2,
+            _ => false,
+        })
+    {
+        return false;
+    };
+    *row = WitnessColumnMap::from(row.values().zip(other.values()).map(|(cell1, cell2)| {
+        match (&cell1.value, &cell2.value) {
+            (CellValue::Known(_), _) => cell1.clone(),
+            _ => cell2.clone(),
+        }
+    }));
+    true
+}
+
+// TODO will be impl method of row later.
+pub fn value_is_known<T: FieldElement>(row: &Row<'_, T>, poly_id: &PolyID) -> bool {
+    row[poly_id].value.is_known()
+}
+
+// TODO will be impl method of row later.
+pub fn set_cell_unknown<T: FieldElement>(row: &mut Row<'_, T>, poly_id: &PolyID) {
+    row[poly_id].value = CellValue::Unknown;
+}
+
+// TODO will be impl method of row later.
+/// Returns true if the values and known constraints are equal for the two rows.
+pub fn rows_are_equal<T: FieldElement>(row1: &Row<'_, T>, row2: &Row<'_, T>) -> bool {
+    row1.values()
+        .zip(row2.values())
+        .all(|(cell1, cell2)| cell1.value == cell2.value)
+}
+
+// TODO will be impl method of row later.
+// TODO we need to get rid of column_ids and ensure that the vector is already in that shape.
+pub fn finalize_row<'a, T: FieldElement>(
+    row: &Row<'_, T>,
+    column_ids: impl IntoIterator<Item = &'a PolyID>,
+) -> (Vec<T>, BitVec) {
```

Comment:
You could introduce a `FinalizedRow` struct here and use it in `FinalizableData`.